File: tests/test_otioz.py
#!/usr/bin/env python
#
# SPDX-License-Identifier: Apache-2.0
# Copyright Contributors to the OpenTimelineIO project

"""Tests for the OTIOZ adapter."""

import unittest
import os
import pathlib
import tempfile

import opentimelineio as otio
import opentimelineio.test_utils as otio_test_utils


class OTIOZTester(unittest.TestCase, otio_test_utils.OTIOAssertions):
    def test_round_trip(self):
        with tempfile.TemporaryDirectory() as temp_dir:

            # Create a timeline
            tl = otio.schema.Timeline()
            tr = otio.schema.Track()
            tl.tracks.append(tr)
            cl = otio.schema.Clip()
            tr.append(cl)

            # Add a media reference
            ref = otio.schema.ExternalReference("video.mov")
            cl.media_reference = ref
            pathlib.Path(os.path.join(temp_dir, ref.target_url)).touch()

            # Write to otioz
            otioz_path = os.path.join(temp_dir, "round_trip.otioz")
            otio.adapters.write_to_file(
                tl,
                otioz_path,
                relative_media_base_dir=temp_dir)

            # Read from otiod
            result = otio.adapters.read_from_file(otioz_path)
            self.assertIsNotNone(result)

    def test_media_policy_as_string(self):
        # As for otiod: the policy may arrive as a string.
        with tempfile.TemporaryDirectory() as temp_dir:
            tl = otio.schema.Timeline()
            tr = otio.schema.Track()
            tl.tracks.append(tr)
            cl = otio.schema.Clip()
            tr.append(cl)
            cl.media_reference = otio.schema.ExternalReference(
                "http://example.com/video.mov")

            otioz_path = os.path.join(temp_dir, "policy.otioz")
            otio.adapters.write_to_file(
                tl,
                otioz_path,
                relative_media_base_dir=temp_dir,
                media_policy="AllMissing")
            self.assertIsNotNone(otio.adapters.read_from_file(otioz_path))


if __name__ == "__main__":
    unittest.main()

File: tests/test_otioz.cpp
// SPDX-License-Identifier: Apache-2.0
// Copyright Contributors to the OpenTimelineIO project

#include "utils.h"

#include <opentimelineio/bundle.h>
#include <opentimelineio/clip.h>
#include <opentimelineio/externalReference.h>
#include <opentimelineio/timeline.h>
#include <opentimelineio/track.h>

#include <fstream>

using namespace OTIO_NS;
using namespace OTIO_NS::bundle;

void create_file(std::filesystem::path const& path)
{
    std::ofstream fs(path);
}

SerializableObject::Retainer<Timeline> create_simple_timeline()
{
    SerializableObject::Retainer<Timeline> tl(new Timeline);

    SerializableObject::Retainer<Track> tr = new Track();
    tl->tracks()->append_child(tr);

    SerializableObject::Retainer<Clip> cl = new Clip();
    tr->append_child(cl);

    return tl;
}

void compare_filenames(Timeline* a, Timeline* b)
{
    const auto a_clips = a->find_clips();
    const auto b_clips = b->find_clips();
    assertEqual(a_clips.size(), b_clips.size());

    for (size_t i = 0; i < a_clips.size(); ++i)
    {
        auto a_ref = dynamic_cast<ExternalReference*>(a_clips[i]->media_reference());
        auto b_ref = dynamic_cast<ExternalReference*>(b_clips[i]->media_reference());
        assertNotNull(a_ref);
        assertNotNull(b_ref);

        assertEqual(
            std::filesystem::u8path(a_ref->target_url()).filename(),
            std::filesystem::u8path(b_ref->target_url()).filename());
    }
}

int
main(int argc, char** argv)
{
    Tests tests;

    tests.add_test("test_otioz_round_trip", [] {
        auto tl = create_simple_timeline();
        SerializableObject::Retainer<ExternalReference> ref =
            new ExternalReference("video.mov");
        tl->find_clips().front()->set_media_reference(ref);
        
        TempDir temp;
        create_file(temp.path() / ref->target_url());
        
        std::string const otioz_path =
            (temp.path() / "round_trip.otioz").u8string();
        WriteOptions options;
        options.relative_media_path = temp.path().u8string();
        assertTrue(write_otioz(tl, otioz_path, options));
        
        auto result = dynamic_cast<Timeline*>(read_otioz(otioz_path));
        assertNotNull(result);
        compare_filenames(tl, result);
    });

    tests.run(argc, argv);
    return 0;
}
